packages/upload-api/src/ucan/conclude.js

```diff
@@ -0,0 +1,163 @@
+import { conclude } from '@web3-storage/capabilities/ucan'
+import { equals } from 'uint8arrays/equals'
+import { DecodeBlockOperationFailed } from '../errors.js'
+
+/**
+ * @param {API.ConcludeServiceContext} context
+ * @returns {API.ServiceMethod<API.UCANConclude, API.UCANConcludeSuccess, API.UCANConcludeFailure>}
+ */
+export const ucanConcludeProvider = ({
+  id,
+  receiptsStorage,
+  tasksStorage,
+  tasksScheduler,
+}) =>
+  provide(conclude, async ({ invocation }) => {
+    const receipt = getConcludeReceipt(invocation)
+
+    // Verify invocation exists failing with ReceiptInvocationNotFound
+    const ranInvocation = receipt.ran
+    const httpPutTaskGetRes = await tasksStorage.get(ranInvocation.link())
+    if (httpPutTaskGetRes.error) {
+      return httpPutTaskGetRes
+    }
+
+    // Store receipt
+    const receiptPutRes = await receiptsStorage.put(receipt)
+    if (receiptPutRes.error) {
+      return {
+        error: receiptPutRes.error,
+      }
+    }
+
+    // THIS IS A TEMPORARY HACK
+    // Schedule `blob/accept` if there is a `http/put` capabilities
+    // inside the invocation that this receipt comes from
+    const scheduleRes = await Promise.all(
+      httpPutTaskGetRes.ok.capabilities
+        // Go through invocation tasks and get all `http/put`
+        .filter((cap) => cap.can === HTTP.put.can)
+        // @ts-expect-error body exists in `http/put` but unknown type here
+        .map(async (/** @type {API.HTTPPut} */ cap) => {
+          // Get triggering task (blob/allocate) by checking blocking task from `url`
+          /** @type {API.UnknownLink} */
+          // @ts-expect-error ts does not know how to get this
+          const blobAllocateTaskCid = cap.nb.url['ucan/await'][1]
+          const blobAllocateTaskGet = await tasksStorage.get(blobAllocateTaskCid)
+          if (blobAllocateTaskGet.error) {
+            return blobAllocateTaskGet
+          }
+
+          /** @type {API.BlobAllocate} */
+          // @ts-expect-error ts does not know how to get this
+          const allocateCapability = blobAllocateTaskGet.ok.capabilities.find(
+            // @ts-expect-error ts does not know how to get this
+            (/** @type {API.BlobAllocate} */ allocateCap) =>
+              equals(allocateCap.nb.blob.digest, cap.nb.body.digest) && allocateCap.can === W3sBlob.allocate.can
+          )
+          
+          const blobAccept = await W3sBlob.accept
+            .invoke({
+              issuer: id,
+              audience: id,
+              with: id.toDIDKey(),
+              nb: {
+                blob: cap.nb.body,
+                exp: Number.MAX_SAFE_INTEGER,
+                space: allocateCapability.nb.space,
+                _put: {
+                  'ucan/await': ['.out.ok', ranInvocation.link()],
+                },
+              },
+              expiration: Infinity,
+            })
+            .delegate()
+
+          return tasksScheduler.schedule(blobAccept)
```

Comment:
I have not read through the scheduler yet, but I do find it surprising that invocation is scheduled here. I would expect this invocation to be already scheduled / pending and here we would just wake it by `blobAllocateTaskCid`.

Reply:
Ok, I think we are not aligned on what the schedule is. For me it is an async execution of the invocation out of this context. So schedule it to be run, can we move forward this way while we do not have the real scheduler?

Reply:
> Ok, I think we are not aligned on what the schedule is. For me it is an async execution of the invocation out of this context. 

I mean the same thing, however I am confused because I would expect that `blobAccept` was already invoked before, so I guess we are rerunning it here ? Which is fine but perhaps adding a comment explaining that would help.

>  So schedule it to be run, can we move forward this way while we do not have the real scheduler?

Yeah, I did not meant to imply we need real scheduler was just confused by what was going on there. Assuming above guess is correct I understand how it works now, but again some code comments would really help to demistify